design/resource-dependency.md

```diff
@@ -0,0 +1,348 @@
+
+### 4. NSQ deployment
+
+An [NSQ](https://github.com/nsqio/nsq) cluster is composed with three components: nsqd, nsqlookup and nsqadmin.
+To set up a `nsq` cluster, one needs to first make sure that the `nsqd` component is up, then start and make sure that the `nsqlookup` component is up, and run the `nsqadmin` component at last.
+nsqlookup consumes the result of nsqd component instance, and nsqadmin consumes the result of nsqlookup component instance.
+If we couldn't describe the dependencies order between the components, the component couldn't run properly.
+
+This is similar to [kubernetes#65502](https://github.com/kubernetes/kubernetes/issues/65502) , the issue describes dependencies between containers on the same Pod, while we are discussing dependencies between components.
+
+### 5. etcd cluster backup
+
+An etcd cluster would be created via EtcdCluster CR, and automatic periodic etcd backup would be made via EtcdBackup CR.
+In OAM, we model EtcdCluster as Component and EtcdBackup as Trait.
+However, the EtcdBackup depends on the connection secret specified in `status.connSecret` field of EtcdCluster.
+In other words, the EtcdBackup trait depends on EtcdCluster component.
+
+## Proposal
+
+The overall idea is to have each resource specify data inputs coming from data outputs (i.e. other resources' fields).
+The OAM runtime will build the dependency graph accordingly and only create dependent resources once the data outputs are ready.
+To achieve this, we propose:
+
+
+1. Add a field `dataOutputs` in Component configuration to specify a data output source:
+
+    ```yaml
+    dataOutputs:
+    - name: <dataoutput-name>
+      fieldPath: <object-field-path>
+    ```
+
+    For example, a my-rds database that outputs connection secret could be specified as:
+
+    ```yaml
+    kind: Component
+    metadata:
+      name: my-rds
+    spec:
+      workload:
+        ... # connection secret name will be written to 'status.connSecret'
+
+    ---
+    kind: ApplicationConfiguration
+    spec:
+      components:
+      - componentName: my-rds
+        dataOutputs:
+        - name: mysql-conn-secret
+          fieldPath: status.connSecret
+    ```
+
+2. Add a field `valueFrom.dataOutputName` in Component ParameterValue to specify the dependency on data outputs:
+
+    ```yaml
+    parameterValues:
+    - name: <parameter-name>
+      valueFrom:
+        dataOutputName: <dataoutput-object-name>
+    ```
+
+    For example, web app's dependency on RDS's database secret could be specified as:
+
+    ```yaml
+    kind: Component
+    metadata:
+      name: my-web
+    spec:
+      workload:
+        kind: MyWorkload
+        metadata:
+          name: my-web
+      parameters:
+      - name: database-secret # workload's fieldPaths needs to be filled
+        fieldPaths:
+        - spec.connSecret
+    ---
+
+    kind: ApplicationConfiguration
+    spec:
+      components:
+      - componentName: my-web
+        parameterValues:
```

Comment:
To echo @ryanzhang-oss 's [comment](https://github.com/crossplane/oam-kubernetes-runtime/pull/24/files#r432096300), I'd propose we put data input/output model alongside with `parameters/parameterValues`, then we will gradually deprecate the later feature in the upcoming releases. In that case, we will have a consistent model for both component and trait in dependency description. My assumption is developer doesn't need to declare `parameters` in component to make data input/output model work. 

But what I'm not sure is: who should declare data input/output? In this proposal, it's operator/system. I have feeling that maybe it should be developer? In that case, we should define a `Application` or `ComponentGroup` first and move data input/output. Let's discuss this part.

Note that, I still think it's a valid case for developer to claim `parameters` as "overridable", but  it's very tricky to guarantee this in implementation (AppConfig controller). I'd suggest we merge this use case into "policies" and enforce it by policy engine (if possible).

Reply:
> But what I'm not sure is: who should declare data input/output? In this proposal, it's operator/system. I have feeling that maybe it should be developer? In that case, we should define a Application or ComponentGroup first and move data input/output. Let's discuss this part.

100% agree this needs more thought. 

I'd make the case that it's the developer - developers already think of applications this way and already solve these problems.  

Ask the question - if I'm the developer - *how do I run my application with dependencies today to test a change?* They are likely writing a mock, using docker-compose, or just launching multiple containers/processes in on their own machine.  For a database, they are likely running one in a container or against a staging instance in the cloud.

You can't get very far as a developer without knowing the *identity* of the other services your component talks to. Most often (based on my customer conversations) the components they communicate with are also written by them/their-team.

It's a big win if we can keep these concepts high level where possible - *"I need to talk to weather-api"* allows us to build a better system than *"I need to declare a parameter to set an environment variable so that someone else with a different job can satisfy a binding from another component..."*.

Reply:
We had a offline but similar discussion around the persona here. So one reasonable approach is:

```yaml
kind: Component
spec:
  workload:
      ...
  dataOutputs:
  - name: etcd-conn-secret
  ...
```

```yaml
kind: ApplicationConfiguration
spec:
  components:
  - componentName: my-etcd-cluster
    traits:
    - trait:
        apiVersion: example.io/v1alpha1
        kind: EtcdBackup
        metadata:
          name: my-etcd-backup
          ...
      dataInputs:
      - from: etcd-conn-secret
        toFieldPaths:
        - spec.connSecret
```

i.e. `Component` and `Trait (AppConfig)` has its own data inputs/outputs claim individually. /cc @hongchaodeng

Reply:
Shouldn't Component's data outputs also sit in ApplicatoinConfiguration:

```
kind: ApplicationConfiguration
spec:
  components:
  - componentName: my-etcd-cluster
    dataOutputs:
      - name: etcd-conn-secret
        ...
    traits:
    - trait:
        apiVersion: example.io/v1alpha1
        kind: EtcdBackup
        metadata:
          name: my-etcd-backup
          ...
      dataInputs:
      - from: etcd-conn-secret
        toFieldPaths:
        - spec.connSecret
```

Reply:
I assume @rynowak 's point is component dependency is knowledge of developers. So the components' data inputs/outputs should be part of `kind:component` object.

Reply:
We can definitely add data inputs in Component definition.
But in this proposal, we would solve the Component configuration part first.
For the use cases described here, it is the operator role that knows the relationship between Components and traits.